src/main/config.cpp

```diff
@@ -0,0 +1,75 @@
+#include <cstdio>
+#include <iostream>
+#include <sstream>
+#include "gflags/gflags.h"
+#include "config.hpp"
+#include "exec_monitor.hpp"
+
+const char DELIMITER = ',';
+
+/*
+	FLAGS SECTION
+*/
+
+// EXEC FLAG - MONITOR PROCESS EXECUTION
+DEFINE_string(ppid, "", "Specifies the parent pid(s) of the processes to monitor using comma-separated values.");
+DEFINE_string(name, "", "Specifies the name(s) of the processes to monitor using comma-separated values.");
+DEFINE_int32(n, 0, "Specifies the number of processes to monitor. The program will stop after n processes were executed.");
+DEFINE_int32(t, 0, "Specifies the number of seconds to monitor. The program will stop after t seconds.");
+
+// DATA EXPORT FLAGS
+DEFINE_string(save, "", "Specifies the format in which to save the terminal output in a separate file. Available formats are csv and protobuf.");
+DEFINE_validator(save, Config::check_save_and_export_format);
+DEFINE_string(export, "", "Specifies the format in which to export data in a separate file. Using export will not print to stdout. Available formats are csv and protobuf.");
+DEFINE_validator(export, Config::check_save_and_export_format);
+
+/*
+	END OF FLAGS SECTION
+*/
+
+Config::Config() {}
+
+int Config::exec_monitor_parse_args_run()
+{
+	if (!gflags::GetCommandLineFlagInfoOrDie("ppid").is_default)
```

Comment:
All the parsing should be done in the constructor.

Reply:
I have one question about this. I imagined that each usecase would have its own flags (for example, exec now could have ppids, names and n_proc, but another usecase can have different flags - even if some flags overlap). So they way I though of this is to first detect the usecase (from --monitor) and then parse the flags, based on the specified usecase (because it does not make sense to try and parse other flags, which are not needed for a particular usecase). For example, although not implemented, let's consider the kernel modules monitoring - it does not need an n_proc flag. Adding all the parsing in the constructor of the config can make this function very long (what is each feature has 3-4 individual flags?) and also, I would check for flags that certainly won't be used for that particular usecase (even if the user said ./hawk --monitor=kmod -nproc 10, I would end up saving n_proc but never using it because it's not a necessary flag for kmod, but for exec.) So, I could parse all the flags in the Config constructor, but for these reasons I thought they should be parsed in separate functions depending on the usecase specified.

Reply:
The constructor can itself call subfunctions. For example, the constructor could do:

```
Config::Config () {
    usecase = find_usecase;
    switch(usecase) {
            case EXEC:
                exec_monitor_parse_args();
                break;
            default:
               break;
    }
}
```

What matters is that when the user of Config creates a config object, it is already ready to be used. And since you want to abstract the complexity of parsing specific subflags from the user, you can make exec_monitor_parse_args private. Then the user never needs to know about the details of your implementation, it is all abstracted.

Reply:
Ok, sounds good :D I was initially thinking of not using the same switch idea in the main function and the constructor as well, but it works.